lib/spells/effects/Catapult.cpp

```diff
@@ -158,10 +238,52 @@ void Catapult::apply(ServerCallback * server, const Mechanics * m, const EffectT
 		server->apply(&removeUnits);
 }
 
+std::vector<EWallPart> Catapult::getPotentialTargets(const Mechanics * m, bool bypassGateCheck, bool bypassTowerCheck) const
+{
+	std::vector<EWallPart> potentialTargets;
+	constexpr std::array<EWallPart, 4> walls = { EWallPart::BOTTOM_WALL, EWallPart::BELOW_GATE, EWallPart::OVER_GATE, EWallPart::UPPER_WALL };
+	constexpr std::array<EWallPart, 3> towers= { EWallPart::BOTTOM_TOWER, EWallPart::KEEP, EWallPart::UPPER_TOWER };
+	constexpr EWallPart gates = EWallPart::GATE;
+
+	// in H3, catapult under automatic control will attack objects in following order:
+	// walls, gates, towers
+	for (auto & part : walls)
+		if (isWallPartAttackable(part, m))
+			potentialTargets.push_back(part);
+
+	if ((potentialTargets.empty() || bypassGateCheck) && (isWallPartAttackable(gates, m)))
+		potentialTargets.push_back(gates);
+
+	if (potentialTargets.empty() || bypassTowerCheck)
+		for (auto & part : towers)
+			if (isWallPartAttackable(part, m))
+				potentialTargets.push_back(part);
+
+	return potentialTargets;
+}
+
+void Catapult::adjustHitChance()
+{
+	keep = vstd::abetween(keep, 0, 100);
+	tower = vstd::abetween(tower, 0, 100);
+	gate = vstd::abetween(gate, 0, 100);
+	wall = vstd::abetween(wall, 0, 100);
+	crit = vstd::abetween(crit, 0, 100);
+	hit = vstd::abetween(hit, 0, 100 - crit);
+	noDmg = vstd::amin(noDmg, 100 - hit - crit);
```

Comment:
`abetween` already assigns value, so `vstd::abetween(keep, 0, 100);` is enough

Reply:
removed assignments